openvdb/openvdb/tools/Morphology.h

```diff
@@ -63,540 +62,831 @@ enum NearestNeighbors { NN_FACE = 6, NN_FACE_EDGE = 18, NN_FACE_EDGE_VERTEX = 26
+            , mOp(op) {}
+
+        /// @brief Dilate a single leaf node by the current spatial scheme
+        ///        stored on the instance of this NodeMaskOp. Neighbour leaf
+        ///        nodes are also updated.
+        /// @details  Unlike erode, dilate is expected to be called in a
+        ///           single threaded context as it will update the node masks
+        ///           of neighbouring leaf nodes as well as the provided leaf.
+        /// @param  leaf  The leaf to dilate. The leaf's origin and value mask
+        ///               are used to calculate the result of the dilation.
+        inline void dilate(LeafType& leaf)
         {
-            if (init) {
-                init = false;
-                Coord orig = xyz.offsetBy(DX*LEAF_DIM, DY*LEAF_DIM, DZ*LEAF_DIM);
-                leaf = acc.probeLeaf(orig);
-                if ((leaf == nullptr) && !acc.isValueOn(orig)) leaf = acc.touchLeaf(orig);
-            }
-            static const int N = (LEAF_DIM - 1)*(DY + DX*LEAF_DIM);
-            if (leaf) leaf->getValueMask().template getWord<Word>(indx-N) |= mask;
+            // copy the mask
+            const MaskType cache = leaf.getValueMask();
+            this->dilate(leaf, cache);
         }
 
-        template<int DX, int DY, int DZ>
-        Word gather(AccessorType& acc, const Coord &xyz, int indx)
+        /// @brief Dilate a single leaf node by the current spatial scheme
+        ///        stored on the instance of this NodeMaskOp. The provided
+        ///        mask is used in place of the actual leaf's node mask and
+        ///        applied to the leaf afterwards. Neighbour leaf nodes are
+        ///        also updated.
+        /// @details  Unlike erode, dilate is expected to be called in a
+        ///           single threaded context as it will update the node masks
+        ///           of neighbouring leaf nodes as well as the provided leaf.
+        /// @param  leaf  The leaf to dilate. The leaf's origin is used to
+        ///               calcualte the result of the dilation.
+        /// @param  mask  The node mask to use in place of the current leaf
+        ///               node mask.
+        inline void dilate(LeafType& leaf, const MaskType& mask)
         {
-            if (init) {
-                init = false;
-                Coord orig = xyz.offsetBy(DX*LEAF_DIM, DY*LEAF_DIM, DZ*LEAF_DIM);
-                leaf = acc.probeLeaf(orig);
-                isOn = leaf ? false : acc.isValueOn(orig);
+            this->clear();
+            mNeighbours[0] = &(leaf.getValueMask());
+            this->setOrigin(leaf.origin());
+            switch (mOp) {
+                case NN_FACE_EDGE        : { this->dilate18(mask); return; }
+                case NN_FACE_EDGE_VERTEX : { this->dilate26(mask); return; }
+                case NN_FACE             : { this->dilate6(mask);  return; }
+                default                  : { this->dilate6(mask);  return; }
```

Comment:
Why not assert(false) and do nothing in this case?

Reply:
Yeah valid point, will change